tools/compute_intersection.cpp

```diff
@@ -117,24 +117,17 @@ int main(int argc, const char** argv)
     IntersectionType intersection_type =
         combinations ? IntersectionType::Combinations : IntersectionType::Query;
 
-    /**/
-    if (false) {
-#define LOOP_BODY(R, DATA, T)                               \
-    }                                                       \
-    else if (app.index_encoding() == BOOST_PP_STRINGIZE(T)) \
-    {                                                       \
-        intersect<BOOST_PP_CAT(T, _index), wand_raw_index>( \
-            app.index_filename(),                           \
-            app.wand_data_path(),                           \
-            filtered_queries,                               \
-            intersection_type,                              \
-            max_term_count);                                \
-        /**/
-
-        BOOST_PP_SEQ_FOR_EACH(LOOP_BODY, _, PISA_INDEX_TYPES);
-#undef LOOP_BODY
-
-    } else {
-        spdlog::error("Unknown type {}", app.index_encoding());
+    try {
+        with_index(app.index_encoding(), app.index_filename(), [&](auto index) {
+            intersect<decltype(index), wand_raw_index>(
+                app.index_filename(),
```

Comment:
I am confused, if we do the mapping in the `with_index` function, why do we pass  the index filename and not the index already mapped?

Reply:
We don't have the index here. This is the whole point, we don't know the type, and we can get away with it because we pass a template function that takes any index. Then it's all handled in `with_index`. It's a workaround to avoid having ifs and elses everywhere we load an index.

Maybe `load_and_run` would be a better name?

I think I mentioned it somewhere on slack, there might be an alternative to write a function that loads an index to `std::variant`. Then we would have to run
```
std::visit([&](auto index) {
    // Do stuff.
}, load_index(...));
```
I don't see much advantage in this, it's essentially the same. It could make a difference if we want to pass the index around, which might be handy.

I actually suggest having both methods. One shortcut for just "run with this index type", and the other "give me an index, I'll handle matching later on my own".

In the latter case, I also suggest having a custom function, hence my naming. I imagined having:
```
load_index(encoding, path);
with_index(index_as_variant, fn);
with_index(encoding, path, fn);
```

Not sure about the name though. Maybe `run_for_index`? Idk, open to suggestions.